File: sdk/golang/tests/robustness/ws/reliability_test.go
package rest

import (
	"fmt"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/extension/interceptor"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/api"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/spot/spotpublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"github.com/stretchr/testify/assert"
	"os"
	"sync"
	"sync/atomic"
	"testing"
	"time"
)

func TestResubscribe(t *testing.T) {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	websocketOptions := types.NewWebSocketClientOptionBuilder().Build()

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		WithWebSocketClientOption(websocketOptions).
		Build()

	client := api.NewClient(option)
	spotWsApi := client.WsService().NewSpotPublicWS()
	err := spotWsApi.Start()
	assert.Nil(t, err)
	defer spotWsApi.Stop()

	btc := atomic.Bool{}
	eth := atomic.Bool{}
	doge := atomic.Bool{}

	cb := func(topic string, subject string, data *spotpublic.TickerEvent) error {
		if topic == "/market/ticker:BTC-USDT" {
			btc.Store(true)
		}

		if topic == "/market/ticker:ETH-USDT" {
			eth.Store(true)
		}

		if topic == "/market/ticker:DOGE-USDT" {
			doge.Store(true)
		}

		fmt.Println(topic)
		return nil
	}

	subid, err := spotWsApi.Ticker([]string{"BTC-USDT", "ETH-USDT"}, cb)
	assert.Nil(t, err)
	subid2, err := spotWsApi.Ticker([]string{"ETH-USDT", "DOGE-USDT"}, cb)
	assert.Nil(t, err)

	time.Sleep(2 * time.Second)

	assert.True(t, btc.Load())
	assert.True(t, eth.Load())
	assert.True(t, doge.Load())

	err = spotWsApi.UnSubscribe(subid2)
	assert.Nil(t, err)
	btc.Store(false)
	eth.Store(false)
	doge.Store(false)

	time.Sleep(2 * time.Second)

	assert.True(t, btc.Load())
	assert.False(t, eth.Load())
	assert.False(t, doge.Load())

	err = spotWsApi.UnSubscribe(subid)
	assert.Nil(t, err)
	btc.Store(false)
	eth.Store(false)
	doge.Store(false)
	time.Sleep(2 * time.Second)

	assert.False(t, btc.Load())
	assert.False(t, eth.Load())
	assert.False(t, doge.Load())
}

func TestResubscribe2(t *testing.T) {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	websocketOptions := types.NewWebSocketClientOptionBuilder().Build()

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		WithWebSocketClientOption(websocketOptions).
		Build()

	client := api.NewClient(option)
	spotWsApi := client.WsService().NewSpotPublicWS()
	err := spotWsApi.Start()
	assert.Nil(t, err)
	defer spotWsApi.Stop()

	cb := func(topic string, subject string, data *spotpublic.TickerEvent) error {
		return nil
	}

	_, err = spotWsApi.Ticker([]string{"BTC-USDT", "ETH-USDT"}, cb)
	assert.Nil(t, err)
	_, err = spotWsApi.Ticker([]string{"BTC-USDT", "ETH-USDT"}, cb)
	assert.NotNil(t, err)
}

func TestStartStop(t *testing.T) {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	websocketOptions := types.NewWebSocketClientOptionBuilder().Build()

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		WithWebSocketClientOption(websocketOptions).
		Build()

	client := api.NewClient(option)
	spotWsApi := client.WsService().NewSpotPublicWS()
	err := spotWsApi.Start()
	assert.Nil(t, err)

	cb := func(topic string, subject string, data *spotpublic.TickerEvent) error {
		fmt.Println(data.ToMap())
		return nil
	}

	_, err = spotWsApi.Ticker([]string{"BTC-USDT", "ETH-USDT"}, cb)
	assert.Nil(t, err)

	time.Sleep(2 * time.Second)
	spotWsApi.Stop()
	time.Sleep(2 * time.Second)
}

func TestConcurrentSubscribe(t *testing.T) {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	websocketOptions := types.NewWebSocketClientOptionBuilder().Build()

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		WithWebSocketClientOption(websocketOptions).
		Build()

	client := api.NewClient(option)
	spotWsApi := client.WsService().NewSpotPublicWS()
	err := spotWsApi.Start()
	assert.Nil(t, err)
	defer spotWsApi.Stop()

	symbol := []string{"BTC-USDT", "ETH-USDT", "DOGE-USDT"}

	t1 := atomic.Int32{}
	t2 := atomic.Int32{}
	t3 := atomic.Int32{}
	t4 := atomic.Int32{}
	t5 := atomic.Int32{}

	go func() {
		_, err = spotWsApi.Ticker(symbol, func(topic string, subject string, data *spotpublic.TickerEvent) error {
			t1.Add(1)
			return nil
		})
		assert.Nil(t, err)
	}()
	go func() {
		_, err = spotWsApi.AllTickers(func(topic string, subject string, data *spotpublic.AllTickersEvent) error {
			t2.Add(1)
			return nil
		})
		assert.Nil(t, err)
	}()

	go func() {
		_, err = spotWsApi.OrderbookLevel1(symbol, func(topic string, subject string, data *spotpublic.OrderbookLevel1Event) error {
			t3.Add(1)
			return nil
		})
		assert.Nil(t, err)
	}()

	go func() {
		_, err = spotWsApi.OrderbookLevel5(symbol, func(topic string, subject string, data *spotpublic.OrderbookLevel5Event) error {
			t4.Add(1)
			return nil
		})
		assert.Nil(t, err)
	}()

	go func() {
		_, err = spotWsApi.OrderbookLevel50(symbol, func(topic string, subject string, data *spotpublic.OrderbookLevel50Event) error {
			t5.Add(1)
			return nil
		})
		assert.Nil(t, err)
	}()

	time.Sleep(2 * time.Second)

	assert.True(t, t1.Load() > 0)
	assert.True(t, t2.Load() > 0)
	assert.True(t, t3.Load() > 0)
	assert.True(t, t4.Load() > 0)
	assert.True(t, t5.Load() > 0)
}

func TestReconnect(t *testing.T) {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	websocketOptions := types.NewWebSocketClientOptionBuilder().WithReconnect(true).WithEventCallback(func(event types.WebSocketEvent, msg string) {
		fmt.Printf("event callback, event:%s, msg: %s", event.String(), msg)
	}).Build()

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		WithWebSocketClientOption(websocketOptions).
		Build()

	client := api.NewClient(option)
	spotWsApi := client.WsService().NewSpotPublicWS()
	err := spotWsApi.Start()
	assert.Nil(t, err)
	defer spotWsApi.Stop()

	_, err = spotWsApi.Trade([]string{"BTC-USDT", "ETH-USDT"}, func(topic string, subject string, data *spotpublic.TradeEvent) error {
		fmt.Println(topic)
		return nil
	})
	assert.Nil(t, err)
	_, err = spotWsApi.Ticker([]string{"BTC-USDT", "ETH-USDT"}, func(topic string, subject string, data *spotpublic.TickerEvent) error {
		fmt.Println(topic)
		return nil
	})
	assert.Nil(t, err)

	time.Sleep(600 * time.Second)
	// disconnect network manually
}

func TestReadBufferFull(t *testing.T) {
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	httpOptionBuilder := types.NewTransportOptionBuilder().
		SetKeepAlive(true).
		AddInterceptors(interceptor.NewLoggingInterceptor(false, defaultLogger))

	set := atomic.Bool{}
	wg := sync.WaitGroup{}
	wg.Add(1)
	websocketOptions := types.NewWebSocketClientOptionBuilder().WithReconnect(true).WithEventCallback(func(event types.WebSocketEvent, msg string) {
		fmt.Printf("event callback, event:%s, msg: %s\n", event.String(), msg)

		if event == types.EventReadBufferFull {
			if set.CompareAndSwap(false, true) {
				wg.Done()
			}
		}

	}).WithReadMessageBuffer(1).Build()

	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithTransportOption(httpOptionBuilder.Build()).
		WithWebSocketClientOption(websocketOptions).
		Build()

	client := api.NewClient(option)
	spotWsApi := client.WsService().NewSpotPublicWS()
	err := spotWsApi.Start()
	assert.Nil(t, err)
	defer spotWsApi.Stop()

	_, err = spotWsApi.Ticker([]string{"BTC-USDT", "ETH-USDT"}, func(topic string, subject string, data *spotpublic.TickerEvent) error {
		time.Sleep(2 * time.Second)
		return nil
	})
	assert.Nil(t, err)

	wg.Wait()
}

File: sdk/python/test/robustness/ws/reliability_test.py
import os
import os
import threading
import unittest
import logging
from time import sleep

from kucoin_universal_sdk.api.client import DefaultClient
from kucoin_universal_sdk.extension.interceptor.logging import LoggingInterceptor
from kucoin_universal_sdk.model.client_option import ClientOptionBuilder
from kucoin_universal_sdk.model.constants import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model.transport_option import TransportOptionBuilder
from kucoin_universal_sdk.model.websocket_option import WebSocketClientOptionBuilder

logging.basicConfig(level=logging.DEBUG)

class ReliabilityTest(unittest.TestCase):

    def setUp(self):
        key = os.getenv("API_KEY")
        secret = os.getenv("API_SECRET")
        passphrase = os.getenv("API_PASSPHRASE")

        # create http transport option
        http_transport_option = (
            TransportOptionBuilder()
            .set_interceptors([LoggingInterceptor()])
            .set_max_connection_per_pool(2)
            .set_max_pool_size(2)
            .build()
        )

        websocket_client_option = WebSocketClientOptionBuilder().build()

        # create client option
        client_option = (
            ClientOptionBuilder()
            .set_key(key)
            .set_secret(secret)
            .set_passphrase(passphrase)
            .set_spot_endpoint(GLOBAL_API_ENDPOINT)
            .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
            .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
            .set_transport_option(http_transport_option)
            .set_websocket_client_option(websocket_client_option)
            .build()
        )

        # create API client
        client = DefaultClient(client_option)
        self.kucoin_ws_api = client.ws_service()
        self.api = self.kucoin_ws_api.new_spot_public_ws()
        self.api.start()

    def tearDown(self):
        self.api.stop()

    def test_resubscribe(self):

        def print_cb(topic, subject, data):
            print(data.to_json())

        sub1 = self.api.ticker(["ETH-USDT", "BTC-USDT"], print_cb)
        sleep(1)
        try:
            sub2 = self.api.ticker(["ETH-USDT", "BTC-USDT"], print_cb)
        except Exception as e:
            self.assertIsNotNone(e)
            print(e.__str__())

    def test_resubscribe_2(self):
        doge = threading.Event()
        btc = threading.Event()
        eth = threading.Event()

        def print_cb(topic, subject, data):
            if topic == "/market/ticker:ETH-USDT":
                eth.set()
            if topic == "/market/ticker:BTC-USDT":
                btc.set()
            if topic == "/market/ticker:DOGE-USDT":
                doge.set()
            print(data.to_json())

        sub1 = self.api.ticker(["ETH-USDT", "BTC-USDT"], print_cb)
        sleep(1)
        sub2 = self.api.ticker(["DOGE-USDT", "BTC-USDT"], print_cb)
        sleep(1)

        self.assertTrue(doge.is_set())
        self.assertTrue(btc.is_set())
        self.assertTrue(eth.is_set())

        self.api.unsubscribe(sub2)
        sleep(1)
        doge.clear()
        sleep(1)
        self.assertTrue(not doge.is_set())

    def test_concurrent_resubscribe(self):
        t1 = threading.Event()
        t2 = threading.Event()
        o1 = threading.Event()
        o2 = threading.Event()
        o3 = threading.Event()

        def print_cb(topic, subject, data):
            if subject == 'trade.ticker':
                t1.set()
            if subject == 'trade.l3match':
                t2.set()
            if subject == 'level1':
                o1.set()
            if subject == 'level2':
                o2.set()
            if subject == 'trade.l2update':
                o3.set()
            print(data.to_json())

        self.api.ticker(["ETH-USDT", "BTC-USDT"], print_cb)
        self.api.trade(["ETH-USDT", "BTC-USDT"], print_cb)
        self.api.orderbook_level1(["ETH-USDT", "BTC-USDT"], print_cb)
        self.api.orderbook_level5(["ETH-USDT", "BTC-USDT"], print_cb)
        self.api.orderbook_level50(["ETH-USDT", "BTC-USDT"], print_cb)

        sleep(10)

        self.assertTrue(t1.is_set())
        self.assertTrue(t2.is_set())
        self.assertTrue(o1.is_set())
        self.assertTrue(o2.is_set())
        self.assertTrue(o3.is_set())

    def test_reconnect(self):
        def print_cb(topic, subject, data):
            print(data.to_json())

        self.api.ticker(["ETH-USDT", "BTC-USDT"], print_cb)

        sleep(100)

    def test_read_buffer_with_event(self):
        key = os.getenv("API_KEY")
        secret = os.getenv("API_SECRET")
        passphrase = os.getenv("API_PASSPHRASE")

        # create http transport option
        http_transport_option = (
            TransportOptionBuilder()
            .set_interceptors([LoggingInterceptor()])
            .set_max_connection_per_pool(2)
            .set_max_pool_size(2)
            .build()
        )

        def event_cb(type, data, msg):
            print(f"event: {str(type)}, {data}, {msg}")
            sleep(1)

        websocket_client_option = WebSocketClientOptionBuilder().with_event_callback(event_cb).with_read_message_buffer(
            1).build()

        # create client option
        client_option = (
            ClientOptionBuilder()
            .set_key(key)
            .set_secret(secret)
            .set_passphrase(passphrase)
            .set_spot_endpoint(GLOBAL_API_ENDPOINT)
            .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
            .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
            .set_transport_option(http_transport_option)
            .set_websocket_client_option(websocket_client_option)
            .build()
        )
        client = DefaultClient(client_option)
        kucoin_ws_api = client.ws_service()
        api = kucoin_ws_api.new_spot_public_ws()
        api.start()

        def print_cb(topic, subject, data):
            print(data.to_json())
            sleep(2)

        api.orderbook_level50(["ETH-USDT", "BTC-USDT"], print_cb)

        sleep(10)
        print('done!')
